core/src/main/java/io/undertow/server/protocol/framed/AbstractFramedChannel.java

```diff
@@ -388,7 +388,7 @@ public synchronized R receive() throws IOException {
             } else if (read == -1 && !hasData) {
                 forceFree = true;
                 readChannelDone = true;
-                lastDataRead();
+                getIoThread().execute(()->{lastDataRead();});
```

Comment:
We can slightly simplify to use a method reference
```suggestion
                getIoThread().execute(this::lastDataRead);
```

Should Http2Channel declare `lastDataRead` volatile?

Reply:
@carterkozak good catch! I'll update the PR

Reply:
@carterkozak as a second thought, I'm closing this PR and sticking with the original one. I think that asynchronously invoking lastDataRead from receive might have collateral effects on receive logic. The reason is that receive uses isLastFrameReceived() to decide if we are at last frame and  Http2Channel.isLastFrameReceived returns lastDataRead. So, by doing it asynchronously we risk a receive missing a true isLastFrameReceived.

Reply:
That makes sense, I agree that the synchronized approach is safer. Thank you for fixing the bug!